README.md
# rtf-parser

This is a general RTF parser.  It takes a text stream and produces a document
object representing the parsed document.  In and of itself, this isn't super
useful but it's the building block for other tools to convert RTF into other
formats.


```js
const parseRTF = require('@iarna/rtf-parser')
const fs = require('fs')

parseRTF.string('{\\rtf1\\ansi\\b hi there\\b0}', (err, doc) => {
  …
})

parseRTF.stream(fs.createReadStream('example.rtf'), (err, doc) => {
  …
})

const parser = parseRTF((err, doc) => {
  …
})
fs.createReadStream('example.rtf').pipe(parser)
```

RTF, unlike HTML, is NOT declarative and is instead a series of commands
that mutate document state.  As such, to accurately convert it you have to
load into something tha tracks that state, then emit chunks of text with
whatever that state was when they were emitted.

RTF, like HTML, allows (mostly) seamless degrading when you don't understand
an element.  As such, while this parser is still quite incomplete it is
already useful

RTF fragments are supported.  `\b hi there\b0` will generate a document with
`hi there` flagged as bold text.

The document returned is of the `RTFDocument` class, see below for details.

## RTF FEATURES SUPPORTED

* The following document code pages:<br>
  437, 737, 775, 850, 852, 853, 855, 857, 858, 860, 861, 863, 865, 866, 869,
  932, 1125, 1250, 1251, 1252, 1253, 1254, 1257
* Unicode characters.
* Non-unicode representations of: non-breaking spaces, soft hyphens and non-breaking hyphens.
* Paragraph alignment: center, justified, left and right
* Style resets.
* Setting the style to "plain".
* Bold, Italic, Underline and Strikethrough.
* Superscript and Subscript.
* first line indent (for indenting the first line of paragraphs)
* indent (for indenting the entire block)
* Fonts
  * The following font character sets:<br>
    ASCII, MacRoman, SHIFT_JIS, CP949, JOHAB, CP936, BIG5, CP1253, CP1254,
    CP1258, CP862, CP1256, CP1257, CP1251, CP874, CP238, CP437
* Colors (foreground and background) 
* Margins
* Text direction

## NOTABLY MISSING

Most notably, stylesheets, list styling and tables are not supported.  List
styling degrades cleanly but tables do not.  There are certainly other
required bits from the spec that are currently ignored.

## CLASSES

### RTFDocument

This is the class you get back from the parse functions.  It has some
document global options and the paragraph objects that make up the document.

* marginLeft, marginRight, marginBottom, marginTop — the margins for this document. These are in
  twips, which are one twentieth of a point.
* content — An array of RTFParagraph objects

### RTFParagraph

* style — An object with paragraph level styling information.
  * firstLineIndent
  * indent
  * align
  * valign

* content — An array of RTFSpan objects

### RTFSpan

* value — a string with the content of this region of text.
* style — an object with the span level styling infomration.
  * font
  * fontSize: In half points
  * bold: boolean
  * italic: boolean
  * underline: boolean
  * strikethrough: boolean
  * foreground: color (an object with red, green and blue values, 0-255)
  * background: color
  * dir (rtl or ltr)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.3.1"></a>
## [1.3.1](https://github.com/iarna/rtf-parser/compare/v1.3.0...v1.3.1) (2019-01-08)


### Bug Fixes

* **interpreter:** avoid finish callback after err ([#14](https://github.com/iarna/rtf-parser/issues/14)) ([2882b17](https://github.com/iarna/rtf-parser/commit/2882b17)), closes [#13](https://github.com/iarna/rtf-parser/issues/13)
* **interpreter:** Unknown RTF command error ([#16](https://github.com/iarna/rtf-parser/issues/16)) ([63314fe](https://github.com/iarna/rtf-parser/commit/63314fe)), closes [#13](https://github.com/iarna/rtf-parser/issues/13)



<a name="1.3.0"></a>
# [1.3.0](https://github.com/iarna/rtf-parser/compare/v1.2.0...v1.3.0) (2018-07-08)


### Bug Fixes

* **nulls:** If a file is null terminated, don't emit the null ([7eee9a2](https://github.com/iarna/rtf-parser/commit/7eee9a2)), closes [#5](https://github.com/iarna/rtf-parser/issues/5)


### Features

* **fragments:** Add RTF fragment support ([a3fb225](https://github.com/iarna/rtf-parser/commit/a3fb225)), closes [#8](https://github.com/iarna/rtf-parser/issues/8)
* **line:** Add support for rtf \line command ([1ebda17](https://github.com/iarna/rtf-parser/commit/1ebda17)), closes [#9](https://github.com/iarna/rtf-parser/issues/9)



<a name="1.2.0"></a>
# [1.2.0](https://github.com/iarna/rtf-parser/compare/v1.1.0...v1.2.0) (2018-07-07)


### Bug Fixes

* **unicode:** Allow for interpretation of signed integers ([e27780e](https://github.com/iarna/rtf-parser/commit/e27780e)), closes [#12](https://github.com/iarna/rtf-parser/issues/12)


### Features

* **cp932:** Add support for code page 932 ([#11](https://github.com/iarna/rtf-parser/issues/11)) ([8dc793f](https://github.com/iarna/rtf-parser/commit/8dc793f))
* **fonttbl:** support groups in fonttbl ([#10](https://github.com/iarna/rtf-parser/issues/10)) ([b6c0f37](https://github.com/iarna/rtf-parser/commit/b6c0f37))


index.js
'use strict'
const RTFParser = require('./rtf-parser.js')
const RTFDocument = require('./rtf-document.js')
const RTFInterpreter = require('./rtf-interpreter.js')

module.exports = parse
parse.string = parseString
parse.stream = parseStream

function parseString (string, cb) {
  parse(cb).end(string)
}

function parseStream (stream, cb) {
  stream.pipe(parse(cb))
}

function parse (cb) {
  let errored = false
  const errorHandler = err => {
    if (errored) return
    errored = true
    parser.unpipe(interpreter)
    interpreter.end()
    cb(err)
  }
  const document = new RTFDocument()
  const parser = new RTFParser()
  parser.once('error', errorHandler)
  const interpreter = new RTFInterpreter(document)
  interpreter.on('error', errorHandler)
  interpreter.on('finish', () => {
    if (!errored) cb(null, document)
  })
  parser.pipe(interpreter)
  return parser
}



rtf-group.js
'use strict'

class RTFGroup {
  constructor (parent) {
    this.parent = parent
    this.content = []
    this.fonts = []
    this.colors = []
    this.style = {}
    this.ignorable = null
  }
  get (name) {
    return this[name] != null ? this[name] : this.parent.get(name)
  }
  getFont (num) {
    return this.fonts[num] != null ? this.fonts[num] : this.parent.getFont(num)
  }
  getColor (num) {
    return this.colors[num] != null ? this.colors[num] : this.parent.getFont(num)
  }
  getStyle (name) {
    if (!name) return Object.assign({}, this.parent.getStyle(), this.style)
    return this.style[name] != null ? this.style[name] : this.parent.getStyle(name)
  }
  resetStyle () {
    this.style = {}
  }
  addContent (node) {
    const defaultFont = this.getFont(0) || {family: 'roman', charset: 'ASCII', name: 'Serif'}
    const defaultColor = this.getColor(0) || {red: 0, blue: 0, green: 0}
    node.style = Object.assign({}, this.getStyle())
    node.style.font = this.getFont(node.style.font)
    node.style.foreground = this.getColor(node.style.foreground)
    node.style.background = this.getColor(node.style.background)
    this.content.push(node)
  }
}

module.exports = RTFGroup


rtf-parser.js
'use strict'
const Transform = require('readable-stream').Transform

class RTFParser extends Transform {
  constructor () {
    super({objectMode: true})
    this.text = ''
    this.controlWord = ''
    this.controlWordParam = ''
    this.hexChar = ''
    this.parserState = this.parseText
    this.char = 0
    this.row = 1
    this.col = 1
  }
  _transform (buf, encoding, done) {
    const text = buf.toString('ascii')
    for (let ii = 0; ii < text.length; ++ii) {
      ++this.char
      if (text[ii] === '\n') {
        ++this.row
        this.col = 1
      } else {
        ++this.col
      }
      this.parserState(text[ii])
    }
    done()
  }
  _flush (done) {
    if (this.text !== '\u0000') this.emitText()
    done()
  }
  parseText (char) {
    if (char === '\\') {
      this.parserState = this.parseEscapes
    } else if (char === '{') {
      this.emitStartGroup()
    } else if (char === '}') {
      this.emitEndGroup()
    } else if (char === '\x0A' || char === '\x0D') {
      // cr/lf are noise chars
    } else {
      this.text += char
    }
  }

  parseEscapes (char) {
    if (char === '\\' || char === '{' || char === '}') {
      this.text += char
      this.parserState = this.parseText
    } else {
      this.parserState = this.parseControlSymbol
      this.parseControlSymbol(char)
    }
  }
  parseControlSymbol (char) {
    if (char === '~') {
      this.text += '\u00a0' // nbsp
      this.parserState = this.parseText
    } else if (char === '-') {
      this.text += '\u00ad' // soft hyphen
    } else if (char === '_') {
      this.text += '\u2011' // non-breaking hyphen
    } else if (char === '*') {
      this.emitIgnorable()
      this.parserState = this.parseText
    } else if (char === "'") {
      this.parserState = this.parseHexChar
    } else if (char === '|') { // formula character
      this.emitFormula()
      this.parserState = this.parseText
    } else if (char === ':') { // subentry in an index entry
      this.emitIndexSubEntry()
      this.parserState = this.parseText
    } else if (char === '\x0a') {
      this.emitEndParagraph()
      this.parserState = this.parseText
    } else if (char === '\x0d') {
      this.emitEndParagraph()
      this.parserState = this.parseText
    } else {
      this.parserState = this.parseControlWord
      this.parseControlWord(char)
    }
  }
  parseHexChar (char) {
    if (/^[A-Fa-f0-9]$/.test(char)) {
      this.hexChar += char
      if (this.hexChar.length >= 2) {
        this.emitHexChar()
        this.parserState = this.parseText
      }
    } else {
      this.emitError(`Invalid character "${char}" in hex literal.`)
      this.parserState = this.parseText
    }
  }
  parseControlWord (char) {
    if (char === ' ') {
      this.emitControlWord()
      this.parserState = this.parseText
    } else if (/^[-\d]$/.test(char)) {
      this.parserState = this.parseControlWordParam
      this.controlWordParam += char
    } else if (/^[A-Za-z]$/.test(char)) {
      this.controlWord += char
    } else {
      this.emitControlWord()
      this.parserState = this.parseText
      this.parseText(char)
    }
  }
  parseControlWordParam (char) {
    if (/^\d$/.test(char)) {
      this.controlWordParam += char
    } else if (char === ' ') {
      this.emitControlWord()
      this.parserState = this.parseText
    } else {
      this.emitControlWord()
      this.parserState = this.parseText
      this.parseText(char)
    }
  }
  emitText () {
    if (this.text === '') return
    this.push({type: 'text', value: this.text, pos: this.char, row: this.row, col: this.col})
    this.text = ''
  }
  emitControlWord () {
    this.emitText()
    if (this.controlWord === '') {
      this.emitError('empty control word')
    } else {
      this.push({
        type: 'control-word',
        value: this.controlWord,
        param: this.controlWordParam !== '' && Number(this.controlWordParam),
        pos: this.char,
        row: this.row,
        col: this.col
      })
    }
    this.controlWord = ''
    this.controlWordParam = ''
  }
  emitStartGroup () {
    this.emitText()
    this.push({type: 'group-start', pos: this.char, row: this.row, col: this.col})
  }
  emitEndGroup () {
    this.emitText()
    this.push({type: 'group-end', pos: this.char, row: this.row, col: this.col})
  }
  emitIgnorable () {
    this.emitText()
    this.push({type: 'ignorable', pos: this.char, row: this.row, col: this.col})
  }
  emitHexChar () {
    this.emitText()
    this.push({type: 'hexchar', value: this.hexChar, pos: this.char, row: this.row, col: this.col})
    this.hexChar = ''
  }
  emitError (message) {
    this.emitText()
    this.push({type: 'error', value: message, row: this.row, col: this.col, char: this.char, stack: new Error().stack})
  }
  emitEndParagraph () {
    this.emitText()
    this.push({type: 'end-paragraph', pos: this.char, row: this.row, col: this.col})
  }
}

module.exports = RTFParser


rtf-document.js
'use strict'
const RTFGroup = require('./rtf-group.js')
const RTFParagraph = require('./rtf-paragraph.js')

class RTFDocument extends RTFGroup {
  constructor () {
    super()
    this.charset = 'ASCII'
    this.ignorable = false
    this.marginLeft = 1800
    this.marginRight = 1800
    this.marginBottom = 1440
    this.marginTop = 1440
    this.style = {
      font: 0,
      fontSize: 24,
      bold: false,
      italic: false,
      underline: false,
      strikethrough: false,
      foreground: null,
      background: null,
      firstLineIndent: 0,
      indent: 0,
      align: 'left',
      valign: 'normal'
    }
  }
  get (name) {
    return this[name]
  }
  getFont (num) {
    return this.fonts[num]
  }
  getColor (num) {
    return this.colors[num]
  }
  getStyle (name) {
    if (!name) return this.style
    return this.style[name]
  }
  addContent (node) {
    if (node instanceof RTFParagraph) {
      while (this.content.length && !(this.content[this.content.length - 1] instanceof RTFParagraph)) {
        node.content.unshift(this.content.pop())
      }
      super.addContent(node)
      if (node.content.length) {
        node.style = Object.assign({}, node.content[0].style)
        node.style.font = this.getFont(node.style.font)
        node.style.foreground = this.getColor(node.style.foreground)
        node.style.background = this.getColor(node.style.background)
      }
    } else {
      super.addContent(node)
    }
  }
}

module.exports = RTFDocument


rtf-interpreter.js
'use strict'
const assert = require('assert')
const util = require('util')

const Writable = require('readable-stream').Writable
const RTFGroup = require('./rtf-group.js')
const RTFParagraph = require('./rtf-paragraph.js')
const RTFSpan = require('./rtf-span.js')
const iconv = require('iconv-lite')

const availableCP = [
  437, 737, 775, 850, 852, 853, 855, 857, 858, 860, 861, 863, 865, 866,
  869, 932, 1125, 1250, 1251, 1252, 1253, 1254, 1257 ]
const codeToCP = {
  0: 'ASCII',
  77: 'MacRoman',
  128: 'SHIFT_JIS',
  129: 'CP949', // Hangul
  130: 'JOHAB',
  134: 'CP936', // GB2312 simplified chinese
  136: 'BIG5',
  161: 'CP1253', // greek
  162: 'CP1254', // turkish
  163: 'CP1258', // vietnamese
  177: 'CP862', // hebrew
  178: 'CP1256', // arabic
  186: 'CP1257',  // baltic
  204: 'CP1251', // russian
  222: 'CP874', // thai
  238: 'CP238', // eastern european
  254: 'CP437' // PC-437
}

class RTFInterpreter extends Writable {
  constructor (document) {
    super({objectMode: true})
    this.doc = document
    this.parserState = this.parseTop
    this.groupStack = []
    this.group = null
    this.once('prefinish', () => this.finisher())
    this.hexStore = []
  }
  _write (cmd, encoding, done) {
    const method = 'cmd$' + cmd.type.replace(/-(.)/g, (_, char) => char.toUpperCase())
    if (this[method]) {
      this[method](cmd)
    } else {
      process.emit('error', `Unknown RTF command ${cmd.type}, tried ${method}`)
    }
    done()
  }
  finisher () {
    while (this.groupStack.length) this.cmd$groupEnd()
    const initialStyle = this.doc.content.length ? this.doc.content[0].style : []
    for (let prop of Object.keys(this.doc.style)) {
      let match = true
      for (let para of this.doc.content) {
        if (initialStyle[prop] !== para.style[prop]) {
          match = false
          break
        }
      }
      if (match) this.doc.style[prop] = initialStyle[prop]
    }
  }
  flushHexStore () {
    if (this.hexStore.length > 0) {
      let hexstr = this.hexStore.map(cmd => cmd.value).join('')
      this.group.addContent(new RTFSpan({
        value: iconv.decode(
          Buffer.from(hexstr, 'hex'), this.group.get('charset'))
      }))
      this.hexStore.splice(0)
    }
  }

  cmd$groupStart () {
    this.flushHexStore()
    if (this.group) this.groupStack.push(this.group)
    this.group = new RTFGroup(this.group || this.doc)
  }
  cmd$ignorable () {
    this.flushHexStore()
    this.group.ignorable = true
  }
  cmd$endParagraph () {
    this.flushHexStore()
    this.group.addContent(new RTFParagraph())
  }
  cmd$groupEnd () {
    this.flushHexStore()
    const endingGroup = this.group
    this.group = this.groupStack.pop()
    const doc = this.group || this.doc
    if (endingGroup instanceof FontTable) {
      doc.fonts = endingGroup.table
    } else if (endingGroup instanceof ColorTable) {
      doc.colors = endingGroup.table
    } else if (endingGroup !== this.doc && !endingGroup.get('ignorable')) {
      for (const item of endingGroup.content) {
        doc.addContent(item)
      }
      process.emit('debug', 'GROUP END', endingGroup.type, endingGroup.get('ignorable'))
    }
  }
  cmd$text (cmd) {
    this.flushHexStore()
    if (!this.group) { // an RTF fragment, missing the {\rtf1 header
      this.group = this.doc
    }
    this.group.addContent(new RTFSpan(cmd))
  }
  cmd$controlWord (cmd) {
    this.flushHexStore()
    if (!this.group.type) this.group.type = cmd.value
    const method = 'ctrl$' + cmd.value.replace(/-(.)/g, (_, char) => char.toUpperCase())
    if (this[method]) {
      this[method](cmd.param)
    } else {
      if (!this.group.get('ignorable')) process.emit('debug', method, cmd.param)
    }
  }
  cmd$hexchar (cmd) {
    this.hexStore.push(cmd)
  }
  cmd$error (cmd) {
    this.emit('error', new Error('Error: ' + cmd.value + (cmd.row && cmd.col ? ' at line ' + cmd.row + ':' + cmd.col : '') + '.'))
  }

  ctrl$rtf () {
    this.group = this.doc
  }

  // new line
  ctrl$line () {
    this.group.addContent(new RTFSpan({ value: '\n' }))
  }

  // alignment
  ctrl$qc () {
    this.group.style.align = 'center'
  }
  ctrl$qj () {
    this.group.style.align = 'justify'
  }
  ctrl$ql () {
    this.group.style.align = 'left'
  }
  ctrl$qr () {
    this.group.style.align = 'right'
  }

  // text direction
  ctrl$rtlch () {
    this.group.style.dir = 'rtl'
  }
  ctrl$ltrch () {
    this.group.style.dir = 'ltr'
  }

  // general style
  ctrl$par () {
    this.group.addContent(new RTFParagraph())
  }
  ctrl$pard () {
    this.group.resetStyle()
  }
  ctrl$plain () {
    this.group.style.fontSize = this.doc.getStyle('fontSize')
    this.group.style.bold = this.doc.getStyle('bold')
    this.group.style.italic = this.doc.getStyle('italic')
    this.group.style.underline = this.doc.getStyle('underline')
  }
  ctrl$b (set) {
    this.group.style.bold = set !== 0
  }
  ctrl$i (set) {
    this.group.style.italic = set !== 0
  }
  ctrl$u (num) {
    var charBuf = Buffer.alloc ? Buffer.alloc(2) : new Buffer(2)
    // RTF, for reasons, represents unicode characters as signed integers
    // thus managing to match literally no one.
    charBuf.writeInt16LE(num, 0)
    this.group.addContent(new RTFSpan({value: iconv.decode(charBuf, 'ucs2')}))
  }
  ctrl$super () {
    this.group.style.valign = 'super'
  }
  ctrl$sub () {
    this.group.style.valign = 'sub'
  }
  ctrl$nosupersub () {
    this.group.style.valign = 'normal'
  }
  ctrl$strike (set) {
    this.group.style.strikethrough = set !== 0
  }
  ctrl$ul (set) {
    this.group.style.underline = set !== 0
  }
  ctrl$ulnone (set) {
    this.group.style.underline = false
  }
  ctrl$fi (value) {
    this.group.style.firstLineIndent = value
  }
  ctrl$cufi (value) {
    this.group.style.firstLineIndent = value * 100
  }
  ctrl$li (value) {
    this.group.style.indent = value
  }
  ctrl$lin (value) {
    this.group.style.indent = value
  }
  ctrl$culi (value) {
    this.group.style.indent = value * 100
  }

// encodings
  ctrl$ansi () {
    this.group.charset = 'ASCII'
  }
  ctrl$mac () {
    this.group.charset = 'MacRoman'
  }
  ctrl$pc () {
    this.group.charset = 'CP437'
  }
  ctrl$pca () {
    this.group.charset = 'CP850'
  }
  ctrl$ansicpg (codepage) {
    if (availableCP.indexOf(codepage) === -1) {
      this.emit('error', new Error('Codepage ' + codepage + ' is not available.'))
    } else {
      this.group.charset = 'CP' + codepage
    }
  }

// fonts
  ctrl$fonttbl () {
    this.group = new FontTable(this.group.parent)
  }
  ctrl$f (num) {
    if (this.group instanceof FontTable) {
      this.group.currentFont = this.group.table[num] = new Font()
    } else if (this.group.parent instanceof FontTable) {
      this.group.parent.currentFont = this.group.parent.table[num] = new Font()
    } else {
      this.group.style.font = num
    }
  }
  ctrl$fnil () {
    if (this.group instanceof FontTable || this.group.parent instanceof FontTable) {
      this.group.get('currentFont').family = 'nil'
    }
  }
  ctrl$froman () {
    if (this.group instanceof FontTable || this.group.parent instanceof FontTable) {
      this.group.get('currentFont').family = 'roman'
    }
  }
  ctrl$fswiss () {
    if (this.group instanceof FontTable || this.group.parent instanceof FontTable) {
      this.group.get('currentFont').family = 'swiss'
    }
  }
  ctrl$fmodern () {
    if (this.group instanceof FontTable || this.group.parent instanceof FontTable) {
      this.group.get('currentFont').family = 'modern'
    }
  }
  ctrl$fscript () {
    if (this.group instanceof FontTable || this.group.parent instanceof FontTable) {
      this.group.get('currentFont').family = 'script'
    }
  }
  ctrl$fdecor () {
    if (this.group instanceof FontTable || this.group.parent instanceof FontTable) {
      this.group.get('currentFont').family = 'decor'
    }
  }
  ctrl$ftech () {
    if (this.group instanceof FontTable || this.group.parent instanceof FontTable) {
      this.group.get('currentFont').family = 'tech'
    }
  }
  ctrl$fbidi () {
    if (this.group instanceof FontTable || this.group.parent instanceof FontTable) {
      this.group.get('currentFont').family = 'bidi'
    }
  }
  ctrl$fcharset (code) {
    if (this.group instanceof FontTable || this.group.parent instanceof FontTable) {
      let charset = null
      if (code === 1) {
        charset = this.group.get('charset')
      } else {
        charset = codeToCP[code]
      }
      if (charset == null) {
        return this.emit('error', new Error('Unsupported charset code #' + code))
      }
      this.group.get('currentFont').charset = charset
    }
  }
  ctrl$fprq (pitch) {
    if (this.group instanceof FontTable || this.group.parent instanceof FontTable) {
      this.group.get('currentFont').pitch = pitch
    }
  }

  // colors
  ctrl$colortbl () {
    this.group = new ColorTable(this.group.parent)
  }
  ctrl$red (value) {
    if (this.group instanceof ColorTable) {
      this.group.red = value
    }
  }
  ctrl$blue (value) {
    if (this.group instanceof ColorTable) {
      this.group.blue = value
    }
  }
  ctrl$green (value) {
    if (this.group instanceof ColorTable) {
      this.group.green = value
    }
  }
  ctrl$cf (value) {
    this.group.style.foreground = value
  }
  ctrl$cb (value) {
    this.group.style.background = value
  }
  ctrl$fs (value) {
    this.group.style.fontSize = value
  }

// margins
  ctrl$margl (value) {
    this.doc.marginLeft = value
  }
  ctrl$margr (value) {
    this.doc.marginRight = value
  }
  ctrl$margt (value) {
    this.doc.marginTop = value
  }
  ctrl$margb (value) {
    this.doc.marginBottom = value
  }

// unsupported (and we need to ignore content)
  ctrl$stylesheet (value) {
    this.group.ignorable = true
  }
  ctrl$info (value) {
    this.group.ignorable = true
  }
  ctrl$mmathPr (value) {
    this.group.ignorable = true
  }
}

class FontTable extends RTFGroup {
  constructor (parent) {
    super(parent)
    this.table = []
    this.currentFont = {family: 'roman', charset: 'ASCII', name: 'Serif'}
  }
  addContent (text) {
    this.currentFont.name += text.value.replace(/;\s*$/, '')
  }
}

class Font {
  constructor () {
    this.family = null
    this.charset = null
    this.name = ''
    this.pitch = 0
  }
}

class ColorTable extends RTFGroup {
  constructor (parent) {
    super(parent)
    this.table = []
    this.red = 0
    this.blue = 0
    this.green = 0
  }
  addContent (text) {
    assert(text.value === ';', 'got: ' + util.inspect(text))
    this.table.push({
      red: this.red,
      blue: this.blue,
      green: this.green
    })
    this.red = 0
    this.blue = 0
    this.green = 0
  }
}

module.exports = RTFInterpreter
